src/nomad_material_processing/solution/schema.py

```diff
@@ -0,0 +1,919 @@
+                    val1 = getattr(combined_components[comparison_key], prop, None)
+                    val2 = getattr(component, prop, None)
+                    if val1 and val2:
+                        setattr(combined_components[comparison_key], prop, val1 + val2)
+                    elif val1:
+                        setattr(combined_components[comparison_key], prop, val1)
+                    elif val2:
+                        setattr(combined_components[comparison_key], prop, val2)
+            else:
+                combined_components[comparison_key] = component.m_copy(deep=True)
+
+        combined_components = list(combined_components.values())
+        combined_components.extend(unprocessed_components)
+
+        return combined_components
+
+    def compute_volume(self, logger: 'BoundLogger' = None) -> None:
+        """
+        Compute the volume of the solution by adding the volumes of its components.
+
+        Args:
+            logger (BoundLogger): A structlog logger.
+        """
+        self.calculated_volume = 0 * ureg('milliliter')
+        for component in self.components:
+            if isinstance(component, (SolutionComponent, SolutionComponentReference)):
+                if component.volume:
+                    self.calculated_volume += component.volume
+
+    def normalize(self, archive: 'EntryArchive', logger: 'BoundLogger') -> None:
+        """
+        Normalize method for the `Solution` section. Compute the total volume of the
+        solution. Populates the solvents and solutes based on the `component_role`. In
+        case of components that are solutions, the quantity of their solvents and solutes
+        is scaled based on their volume used. Finally, combines the components with the
+        same CAS number.
+
+        Args:
+            archive (EntryArchive): A NOMAD archive.
+            logger (BoundLogger): A structlog logger.
+        """
+        self.solvents, self.solutes = [], []
+
+        self.compute_volume(logger)
+        volume = self.calculated_volume
+        if self.measured_volume:
+            volume = self.measured_volume
+
+        for component in self.components:
+            if isinstance(component, SolutionComponent):
+                component.mass_fraction = None
+                component.compute_molar_concentration(volume, logger)
+                if component.component_role == 'Solvent':
+                    self.solvents.append(component.m_copy(deep=True))
+                elif component.component_role == 'Solute':
+                    self.solutes.append(component.m_copy(deep=True))
+            elif isinstance(component, SolutionComponentReference):
+                # add solutes and solvents from the solution
+                # while taking the volume used into account
+                if component.system:
+                    scaler = 1
+                    if component.volume:
+                        # update scaler based on the volume used
```

Comment:
mass of the added solution is a non-editable field. The user has to give the volume. We can change this in future if some user wants to add solution with mass instead of volume.

Reply:
Ah, okay. Yes, this should definitely be added in the future but let's skip it for now.